src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1605,215 @@ class OptimizeShuffles : public IRMutator {
 public:
     OptimizeShuffles(int lut_alignment) : lut_alignment(lut_alignment) {}
 };
+
+class VtmpyGenerator : public IRMutator {
+private:
+    using IRMutator::visit;
+    Scope<Expr> vars;
+
+    // Return value of variable if expr is a variable.
+    // Else return the expr.
+    Expr get_val(Expr e) {
+        const Variable *op = e.as<Variable>();
+        if (op) {
+            if (vars.contains(op->name)) {
+                return vars.get(op->name);
+            }
+            else {
+                return Expr();
+            }
+        } else {
+            return e;
+        }
+    }
+
+    // Save variable value
+    void visit(const Let* op) {
+        vars.push(op->name, op->value);
+        IRMutator::visit(op);
+        vars.pop(op->name);
+    }
+
+    // Save variable value
+    void visit(const LetStmt *op) {
+        vars.push(op->name, op->value);
+        IRMutator::visit(op);
+        vars.pop(op->name);
+    }
+
+    // Check if vector start indices differ by constant diff.
+    bool cmp_basediff_const(const Expr &a, const Expr &b, int diff) {
+        Expr maybe_load_a = calc_load(a);
+        Expr maybe_load_b = calc_load(b);
+
+        if (maybe_load_a.defined() && maybe_load_b.defined()) {
+            const Load* load_a = maybe_load_a.as<Load>();
+            const Load* load_b = maybe_load_b.as<Load>();
+            const Ramp* ramp_a = get_val(load_a->index).as<Ramp>();
+            const Ramp* ramp_b = get_val(load_b->index).as<Ramp>();
+            if (ramp_a && ramp_b && load_a->name == load_b->name) {
+                Expr base_diff = simplify(ramp_a->base - ramp_b->base - diff);
+                if (is_const(base_diff, 0)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    // Return the ramp of first vector if all vector in exprs are continuous
+    // vectors in memory.
+    Expr are_continuous_vectors(const vector<Expr> exprs) {
```

Comment:
I see the problem you are concerned about now, it's different than I was thinking at first (a let expression appearing in front of the shuffle without the shuffle simplifying away).

Yes, the shuffle approach isn't necessarily completely robust in this case. On the other hand, the existing code is not robust to doing stencils on some other computed expression, and I think it *might* work with the shuffle approach. So, there are tradeoffs here...